android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java

```diff
@@ -0,0 +1,143 @@
+                .fromCallable(() -> persist(filename, archive));
+
+        Single<UploadSession> uploadSessionSingle = uploadFileSingle
+                .flatMap(this::requestUploadSession);
+
+        return Single.zip(uploadFileSingle, uploadSessionSingle, this::upload)
+                .flatMap(i -> i)
+                .toCompletable();
+    }
+
+    public Single<UploadSession> upload(UploadFile uploadFile, UploadSession session) {
+
+        File file = new File(uploadFile.filename);
+        RequestBody requestBody = RequestBody.create(MediaType.parse(uploadFile.contentType), file);
+
+        return RxUtils.toBodySingle(
+                s3Service.uploadToS3(
+                        session.getUrl(),
+                        requestBody,
+                        uploadFile.md5Hash,
+                        uploadFile.contentType))
+                .toCompletable()
+                .doOnCompleted(() -> {
+                    // TODO: update upload status in DB, delete file
+                    LOG.info("Upload succeeded for id: " + session.getId());
+                }).andThen(
+                        RxUtils.toBodySingle(api.completeUploadSession(session.getId())
+                        ).onErrorReturn((t) -> {
+                            LOG.info("Failed to call upload complete, server will recover", t);
+                            return session;
+                        }));
+    }
+
+    public Single<UploadSession> requestUploadSession(UploadFile uploadFile) {
+        return RxUtils.toBodySingle(
+                api.requestUploadSession(
+                        new UploadRequest()
+                                .name(uploadFile.filename)
+                                .contentType(uploadFile.contentType)
+                                .contentLength(uploadFile.fileLength)
+                                .contentMd5(uploadFile.md5Hash)))
+                .doOnSuccess((uploadSession) -> {
+                    // TODO: insert upload session to DB, link to file metadata
+                    LOG.info("Received upload session with id: " + uploadSession.getId());
+                });
+    }
+
+    public UploadFile persist(String filename, Archive archive) throws IOException,
+            CMSException, NoSuchAlgorithmException {
+        File file = new File(filename);
+
+        ByteSink sink = Files.asByteSink(file, FileWriteMode.APPEND);
+
+        MessageDigest md5 = null;
+        try {
+            md5 = MessageDigest.getInstance("MD5");
+        } catch (NoSuchAlgorithmException e) {
+            LOG.error("MD5 digest not found", e);
+            throw e;
+        }
+
+        DigestOutputStream md5OutStream = new DigestOutputStream(sink.openBufferedStream(), md5);
+
+        OutputStream encryptedOutputStream = encryptor.encrypt(md5OutStream);
+
+        archive.writeTo(encryptedOutputStream);
+
+        String md5Hash = BaseEncoding.base64().encode(md5.digest());
+
+        //TODO: write file metadata to db
```

Comment:
My next PR will store things in SharedPreferences, that's for android-sdk.

That is a good point though, the researchstack-sdk should probably use StorageAccess instead of SharedPreferences directly. I'll include that.